### PR TITLE
kube-apiserver/cmd: stratify construction to follow options/config/server pattern

### PR DESCRIPTION
**Pull request by sttts**
We follow that pattern of options/config/server in all of our apiservers in the delegation chain, but somehow never finished that work in kube-apiserver. This will make kube-apiserver play well in compositions of similar components.

/kind cleanup

```release-note
NONE
```

This is a preparation for a proper generic reusable controlplane package, compare KEP https://github.com/kubernetes/enhancements/pull/4052.

The direction is to turn pkg/controlplane step-by-step into a generic-controlplane package, and eventually move that into a staging repo.

For review, go commmit by commit. They are prefixed with what they do. MOVEs are trivial code moves.

### CODE REVIEW

**Comment by sttts** [comment]
> (smells like a flake!)

Probably real. Have to check.

**Comment by sttts** [comment]
> /retest
> 
> failure in:
> 
> ```
> k8s.io/kubernetes/test/integration/examples: TestAPIServiceWaitOnStart
> ```
> 
> (smells like a flake!)

@dims compare https://github.com/kubernetes/kubernetes/compare/3f7f1b6393180bd0ee56ad414c4fad0f0dcb8626..dbf2d64c2424b41721bf71c5c3114d780c650a0e#diff-170c324cc9e649c56eba833146296ea83b4fe433cb1c05b63ca18578609da4d2R188. Typo accessing the wrong resource config for the CRD group version. Hence, the aggregator thought CRDs are off.

**Comment by sttts** [comment]
Looks like unrelated flake. Nothing changed in that (low-level) area.

/retest

**Comment by dims** [comment]
/hold

(@sttts do you need lgtm from anyone else?)

**Comment by sttts** [comment]
> do you need lgtm from anyone else?

Wouldn't mind another pair of eyes from somebody with apiserver background. Have pinged @p0lyn0mial, or maybe @tkashem.

File: pkg/controlplane/apiserver/apiextensions.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	"k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/apiserver/pkg/util/webhook"
+	"k8s.io/client-go/informers"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+)
+
+func CreateAPIExtensionsConfig(
+	kubeAPIServerConfig server.Config,
+	externalInformers informers.SharedInformerFactory,
+	pluginInitializers []admission.PluginInitializer,
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/controlplane/apiserver/apiextensions.go:38]: to be removed in a follow-up

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/controlplane/apiserver/apiextensions.go:38]: done

File: pkg/controlplane/apiserver/apiextensions.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	"k8s.io/apiserver/pkg/server"
+	"k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/apiserver/pkg/util/webhook"
+	"k8s.io/client-go/informers"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+)
+
+func CreateAPIExtensionsConfig(
+	kubeAPIServerConfig server.Config,
+	externalInformers informers.SharedInformerFactory,
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/controlplane/apiserver/apiextensions.go:37]: Long ago we had internal informers too. Renaming.

File: pkg/controlplane/apiserver/apiextensions.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	pluginInitializers []admission.PluginInitializer,
+	commandOptions *options.ServerRunOptions,
+	masterCount int,
+	serviceResolver webhook.ServiceResolver,
+	authResolverWrapper webhook.AuthenticationInfoResolverWrapper,
+) (*apiserver.Config, error) {
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/controlplane/apiserver/apiextensions.go]: done

**Comment by sttts** [comment]
Rebased.

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -328,182 +307,44 @@ func CreateKubeAPIServerConfig(s completedServerRunOptions) (
...
-		genericConfig.DisabledPostStartHooks.Insert(rbacrest.PostStartHookName)
-	}
+var testServiceResolver webhook.ServiceResolver
 
-	lastErr = s.Audit.ApplyTo(genericConfig)
-	if lastErr != nil {
-		return
+// SetServiceResolverForTests allows the service resolver to be overridden during tests.
+// Tests using this function must run serially as this function is not safe to call concurrently with server start.
+func SetServiceResolverForTests(resolver webhook.ServiceResolver) func() {
```
Comment (p0lyn0mial) [inline cmd/kube-apiserver/app/server.go:357]: shouldn't we move this function to a test pkg?

**Pull request merged**